typescript/sdk/src/fee/EvmTokenFeeModule.ts

```diff
@@ -141,23 +181,117 @@ export class EvmTokenFeeModule extends HyperlaneModule<
+    const address = params?.address ?? this.args.addresses.deployedFee;
+    const routingDestinations = params?.routingDestinations ?? [];
 
-  async read(routingDestinations?: number[]): Promise<DerivedTokenFeeConfig> {
-    return this.reader.deriveTokenFeeConfig(
-      this.args.addresses.deployedFee,
+    return this.reader.deriveTokenFeeConfig({
+      address: address,
       routingDestinations,
-    );
+    });
   }
 
   async update(
-    _config: TokenFeeConfigInput,
-  ): Promise<
-    Annotated<ProtocolTypedTransaction<ProtocolType.Ethereum>['transaction']>[]
-  > {
-    throw new Error('Not implemented');
+    targetConfig: TokenFeeConfigInput,
+    params?: Partial<ReaderParams>,
+  ): Promise<AnnotatedEV5Transaction[]> {
+    let updateTransactions: AnnotatedEV5Transaction[] = [];
+
+    const normalizedTargetConfig: TokenFeeConfig = normalizeConfig(
+      await EvmTokenFeeModule.expandConfig({
+        config: targetConfig,
+        multiProvider: this.multiProvider,
+        chainName: this.chainName,
+      }),
+    );
+
+    const actualConfig = await this.read(params);
+    const normalizedActualConfig: TokenFeeConfig =
+      normalizeConfig(actualConfig);
+
+    //If configs are the same, return empty array
+    if (deepEquals(normalizedActualConfig, normalizedTargetConfig)) {
+      this.logger.debug(
+        `Same config for ${normalizedTargetConfig.type}, no update needed`,
+      );
+      return [];
+    }
+
+    // if the type is a immutable type, then redeploy
+    if (
+      ImmutableTokenFeeType.includes(
+        normalizedTargetConfig.type as (typeof ImmutableTokenFeeType)[number],
+      )
+    ) {
+      this.logger.info(
+        `Immutable fee type ${normalizedTargetConfig.type}, redeploying`,
+      );
+      const contracts = await this.deploy({
+        config: normalizedTargetConfig,
+        multiProvider: this.multiProvider,
+        chainName: this.chainName,
+        contractVerifier: this.contractVerifier,
+      });
+      this.args.addresses.deployedFee =
+        contracts[this.chainName][normalizedTargetConfig.type].address;
+    }
+
+    // if the type is a mutable type, then update
+    updateTransactions = [
+      ...(await this.updateRoutingFee(
+        _.merge(
+          actualConfig,
+          normalizedTargetConfig,
+        ) as DerivedRoutingFeeConfig,
+      )),
+      ...this.createOwnershipUpdateTxs(
+        normalizedActualConfig,
+        normalizedTargetConfig,
+      ),
+    ];
+
+    return updateTransactions;
+  }
+
+  private async updateRoutingFee(targetConfig: DerivedRoutingFeeConfig) {
+    const updateTransactions: AnnotatedEV5Transaction[] = [];
+
+    if (!targetConfig.feeContracts) return [];
+    const currentRoutingAddress = this.args.addresses.deployedFee;
+    await promiseObjAll(
+      objMap(targetConfig.feeContracts, async (chainName, config) => {
+        const address = config.address;
+        await this.update(config, { address });
```

Comment:
I think that this might be problematic as at the end of all the update runs the current module won't be linked to the routing fee module but one of the sub fee contracts instead as `this. actualConfig` will be the result of reading the sub contract. Can't we create a module for each one of the contracts instead or maybe create an internal update method that does not operate on `this`?

Reply:
updated to initialize a new FeeModule for each subFee